fix: `@import` is broken in `0.21.7`

**Description**

`@import` is broken in `0.21.7`.

**Steps To Reproduce**
1. set `jaspr` and `jaspr_builder` version to `0.21.7`.
2. add the same web specific import in more than one file  :
```dart
@Import.onWeb('package:dio/browser.dart', show: [#BrowserHttpClientAdapter])
```
3. run  `jaspr run`.


**Expected Behavior**

app runs.

**actual Behavior**

<img width="1240" height="416" alt="Image" src="https://github.com/user-attachments/assets/07cf063a-25bb-4cab-8de4-71578cffa557" />

```bash
[BUILDER] [ERROR] org-dartlang-app:///packages/frontend/generated/imports/_web.dart:15:9: Error: 'BrowserHttpClientAdapterOrStubbed' is already declared in this scope.
[BUILDER] [ERROR] typedef BrowserHttpClientAdapterOrStubbed = BrowserHttpClientAdapter;
[BUILDER] [ERROR] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[BUILDER] [ERROR] org-dartlang-app:///packages/frontend/generated/imports/_web.dart:14:9: Context: Previous declaration of 'BrowserHttpClientAdapterOrStubbed'.
[BUILDER] [ERROR] typedef BrowserHttpClientAdapterOrStubbed = BrowserHttpClientAdapter;
[BUILDER] [ERROR] ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[BUILDER] [ERROR] jaspr_web_compilers:ddc (lazy) on web/main.clients.bootstrap.ddc.module:
[BUILDER] [ERROR] AssetNotFoundException: frontend|lib/generated/imports/_web.ddc.dill
[BUILDER] [WARNING] jaspr_web_compilers:entrypoint on web/main.clients.bootstrap.dart:
[BUILDER] [WARNING] Unable to read frontend|web/main.clients.bootstrap.ddc.js, check your console or the `.dart_tool/build/generated/frontend/web/main.clients.bootstrap.ddc.js.errors` log file.
```

**Additional Context**

bug disappears when setting `jaspr` and `jaspr_builder` to version `0.21.6`.

--- a/packages/jaspr_builder/lib/src/imports/imports_builder.dart
+++ b/packages/jaspr_builder/lib/src/imports/imports_builder.dart
@@ -2,6 +2,7 @@ import 'dart:async';
 import 'dart:convert';
 
 import 'package:build/build.dart';
+import 'package:collection/collection.dart';
 import 'package:path/path.dart' as p;
 
 import '../utils.dart';
@@ -17,7 +18,8 @@ class ImportsOutputBuilder implements Builder {
       final entries = (jsonDecode(await buildStep.readAsString(buildStep.inputId)) as List<Object?>)
           .cast<Map<String, Object?>>()
           .map(ImportEntry.fromJson)
-          .toList();
+          .toList()
+          .sortedBy((e) => e.url);
       final outputId = buildStep.inputId.changeExtension('.dart');
 
       final webShow = <String>{};

--- a/packages/jaspr_builder/lib/src/imports/stubs_builder.dart
+++ b/packages/jaspr_builder/lib/src/imports/stubs_builder.dart
@@ -2,6 +2,7 @@ import 'dart:async';
 import 'dart:convert';
 
 import 'package:build/build.dart';
+import 'package:collection/collection.dart';
 import 'package:glob/glob.dart';
 import 'package:path/path.dart' as p;
 
@@ -16,67 +17,86 @@ class ImportsStubsBuilder implements Builder {
   FutureOr<void> build(BuildStep buildStep) async {
     final imports = buildStep.findAssets(Glob('**/*.imports.json'));
 
-    final webImports = <String, Set<ImportElement>>{};
-    final vmImports = <String, Set<ImportElement>>{};
+    final webImports = PlatformImports();
+    final vmImports = PlatformImports();
 
-    final stubs = <String>{};
+    final stubs = <String>[];
+    final stubbedNames = <String>{};
+
+    final allEntries = <ImportEntry>[];
 
     await for (final id in imports) {
-      final entries = (jsonDecode(await buildStep.readAsString(id)) as List<Object?>)
-          .cast<Map<String, Object?>>()
-          .map(ImportEntry.fromJson)
-          .toList();
+      final entries = (jsonDecode(await buildStep.readAsString(id)) as List<Object?>).cast<Map<String, Object?>>().map(
+        ImportEntry.fromJson,
+      );
+
       for (final entry in entries) {
-        var url = entry.url;
-
-        for (final elem in entry.elements) {
-          if (elem.type == ElementType.variable) {
-            stubs.add('dynamic ${elem.name};');
-          } else if (elem.type == ElementType.type) {
-            stubs.add('dynamic ${elem.name};');
-            stubs.add('typedef ${elem.name}OrStubbed = dynamic;');
-          } else if (elem.type == ElementType.extension) {
-            stubs.add(
-              'extension ${elem.name} on dynamic {\n'
-              '${elem.details.join('\n')}\n'
-              '}',
-            );
-          }
-        }
-
-        var uri = Uri.parse(url);
+        final uri = Uri.parse(entry.url);
         if (uri.scheme.isEmpty && path.isRelative(uri.path)) {
-          var absUrl = path.join(path.dirname(id.path), url);
-          url = path.relative(absUrl, from: 'lib/generated/imports/_.dart');
-        }
-
-        if (entry.platform == 0) {
-          (webImports[url] ??= {}).addAll(entry.elements);
+          final absUrl = path.join(path.dirname(id.path), entry.url);
+          allEntries.add(
+            ImportEntry(
+              path.relative(absUrl, from: 'lib/generated/imports/_.dart'),
+              entry.platform,
+              entry.elements,
+            ),
+          );
         } else {
-          (vmImports[url] ??= {}).addAll(entry.elements);
+          allEntries.add(entry);
         }
       }
     }
 
-    if (vmImports.isNotEmpty) {
+    allEntries.sortBy((e) => e.url);
+
+    for (final entry in allEntries) {
+      if (entry.platform == 0) {
+        webImports.add(entry.url, entry.elements);
+      } else {
+        vmImports.add(entry.url, entry.elements);
+      }
+
+      for (final elem in entry.elements) {
+        if (stubbedNames.contains(elem.name)) {
+          continue;
+        }
+        stubbedNames.add(elem.name);
+        if (elem.type == ElementType.variable) {
+          stubs.add('dynamic ${elem.name};');
+        } else if (elem.type == ElementType.type) {
+          stubs.add(
+            'dynamic ${elem.name};\n'
+            'typedef ${elem.name}OrStubbed = dynamic;',
+          );
+        } else if (elem.type == ElementType.extension) {
+          stubs.add(
+            'extension ${elem.name} on dynamic {\n'
+            '${elem.details.join('\n')}\n'
+            '}',
+          );
+        }
+      }
+    }
+
+    if (!vmImports.isEmpty) {
       await buildStep.writeAsFormattedDart(AssetId(buildStep.inputId.package, 'lib/generated/imports/_vm.dart'), """
         // ignore_for_file: directives_ordering, deprecated_member_use
         
-        ${vmImports.entries.where((e) => e.value.any((e) => e.type == ElementType.type)).map((e) => "import '${e.key}' show ${e.value.where((e) => e.type == ElementType.type).map((e) => e.name).join(', ')};").join('\n')}
-        ${vmImports.entries.map((e) => "export '${e.key}' show ${e.value.map((e) => e.name).join(', ')};").join('\n')}
+        ${vmImports.imports}
+        ${vmImports.exports}
 
-        ${vmImports.values.expand((v) => v.where((e) => e.type == ElementType.type)).map((e) => 'typedef ${e.name}OrStubbed = ${e.name};').join('\n')}
+        ${vmImports.stubs}
       """);
     }
 
-    if (webImports.isNotEmpty) {
+    if (!webImports.isEmpty) {
       await buildStep.writeAsFormattedDart(AssetId(buildStep.inputId.package, 'lib/generated/imports/_web.dart'), """
         // ignore_for_file: directives_ordering, deprecated_member_use
 
-        ${webImports.entries.where((e) => e.value.any((v) => v.type == ElementType.type)).map((e) => "import '${e.key}' show ${e.value.where((v) => v.type == ElementType.type).map((v) => v.name).join(', ')};").join('\n')}
-        ${webImports.entries.map((e) => "export '${e.key}' show ${e.value.map((e) => e.name).join(', ')};").join('\n')}
+        ${webImports.imports}
+        ${webImports.exports}
 
-        ${webImports.values.expand((v) => v.where((e) => e.type == ElementType.type)).map((e) => 'typedef ${e.name}OrStubbed = ${e.name};').join('\n')}
+        ${webImports.stubs}
       """);
     }
 
@@ -97,4 +117,45 @@ class ImportsStubsBuilder implements Builder {
       'lib/generated/imports/_vm.dart',
     ],
   };
+}
+
+class PlatformImports {
+  final Map<String, String> _uniqueNames = {};
+  final Map<String, Set<String>> _importedNames = {};
+  final Map<String, Set<String>> _exportedNames = {};
+  final Set<String> _stubbedNames = {};
+
+  void add(String url, List<ImportElement> elements) {
+    final importedNames = _importedNames[url] ??= <String>{};
+    final exportedNames = _exportedNames[url] ??= <String>{};
+
+    for (final element in elements) {
+      if (_uniqueNames[element.name] case final existingImport? when existingImport != url) {
+        throw Exception(
+          "Cannot import ${element.name} from $url, because it is also imported from $existingImport. "
+          'Names imported via @Import must be unique for each platform across the project.',
+        );
+      }
+      _uniqueNames[element.name] = url;
+      exportedNames.add(element.name);
+      if (element.type == ElementType.type) {
+        importedNames.add(element.name);
+        _stubbedNames.add(element.name);
+      }
+    }
+  }
+
+  bool get isEmpty => _uniqueNames.isEmpty;
+
+  String get imports => [
+    for (final MapEntry(key: url, value: names) in _importedNames.entries)
+      if (names.isNotEmpty) "import '$url' show ${names.join(', ')};",
+  ].join('\n');
+  String get exports => [
+    for (final MapEntry(key: url, value: names) in _exportedNames.entries)
+      if (names.isNotEmpty) "export '$url' show ${names.join(', ')};",
+  ].join('\n');
+  String get stubs => [
+    for (final name in _stubbedNames.toList()..sort()) "typedef ${name}OrStubbed = $name;",
+  ].join('\n');
 }

--- a/packages/jaspr_builder/test/imports/import_annotation_test.dart
+++ b/packages/jaspr_builder/test/imports/import_annotation_test.dart
@@ -42,5 +42,28 @@ void main() {
         readerWriter: reader,
       );
     });
+
+    test('throws on conflicting names', () async {
+      final result = await testBuilder(
+        ImportsStubsBuilder(BuilderOptions({})),
+        {
+          ...importsSources,
+          ...conflictingImportsSources,
+          ...importsModuleOutput,
+          ...conflictingImportsModuleOutput,
+        },
+        readerWriter: reader,
+      );
+
+      expect(result.succeeded, isFalse);
+      expect(result.errors, hasLength(1));
+      expect(
+        result.errors.first,
+        equals(
+          'Exception: Cannot import Document from package:jaspr/browser.dart, because it is also imported from dart:html. '
+          'Names imported via @Import must be unique for each platform across the project.',
+        ),
+      );
+    });
   });
 }

--- a/packages/jaspr_builder/test/imports/sources/imports.dart
+++ b/packages/jaspr_builder/test/imports/sources/imports.dart
@@ -9,6 +9,22 @@ const importsSources = {
     @Import.onServer('dart:collection', show: [#IterableExtensions])
     import 'target.imports.dart';
   ''',
+  'site|lib/other.dart': '''
+    import 'package:jaspr/jaspr.dart';
+    
+    @Import.onWeb('dart:html', show: [#window, #Window, #Document])
+    @Import.onServer('dart:io', show: [#HttpServer, #HttpHeaders])
+    import 'other.imports.dart';
+  ''',
+};
+
+const conflictingImportsSources = {
+  'site|lib/conflict.dart': '''
+    import 'package:jaspr/jaspr.dart';
+    
+    @Import.onWeb('package:jaspr/browser.dart', show: [#Document])
+    import 'conflict.imports.dart';
+  ''',
 };
 
 final importsModuleOutput = {
@@ -46,6 +62,37 @@ final importsModuleOutput = {
       ],
     },
   ]),
+  'site|lib/other.imports.json': jsonEncode([
+    {
+      "url": "dart:html",
+      "platform": 0,
+      "elements": [
+        {"name": "window", "type": 2, "details": <String>[]},
+        {"name": "Window", "type": 0, "details": <String>[]},
+        {"name": "Document", "type": 0, "details": <String>[]},
+      ],
+    },
+    {
+      "url": "dart:io",
+      "platform": 1,
+      "elements": [
+        {"name": "HttpServer", "type": 0, "details": <String>[]},
+        {"name": "HttpHeaders", "type": 0, "details": <String>[]},
+      ],
+    },
+  ]),
+};
+
+final conflictingImportsModuleOutput = {
+  'site|lib/conflict.imports.json': jsonEncode([
+    {
+      "url": "package:jaspr/browser.dart",
+      "platform": 0,
+      "elements": [
+        {"name": "Document", "type": 0, "details": <String>[]},
+      ],
+    },
+  ]),
 };
 
 final importsOutput = {
@@ -62,7 +109,21 @@ final importsOutput = {
       '\n'
       'export \'generated/imports/_vm.dart\'\n'
       '    if (dart.library.js_interop) \'generated/imports/_stubs.dart\'\n'
-      '    show HttpServer, HttpServerOrStubbed, IterableExtensions;\n'
+      '    show IterableExtensions, HttpServer, HttpServerOrStubbed;\n'
+      '',
+  'site|lib/other.imports.dart':
+      '// dart format off\n'
+      '// ignore_for_file: type=lint\n\n'
+      '// GENERATED FILE, DO NOT MODIFY\n'
+      '// Generated with jaspr_builder\n'
+      '\n'
+      'export \'generated/imports/_web.dart\'\n'
+      '    if (dart.library.io) \'generated/imports/_stubs.dart\'\n'
+      '    show window, Window, WindowOrStubbed, Document, DocumentOrStubbed;\n'
+      '\n'
+      'export \'generated/imports/_vm.dart\'\n'
+      '    if (dart.library.js_interop) \'generated/imports/_stubs.dart\'\n'
+      '    show HttpServer, HttpServerOrStubbed, HttpHeaders, HttpHeadersOrStubbed;\n'
       '',
 };
 
@@ -75,9 +136,10 @@ final importsStubsOutput = {
       '\n'
       '// ignore_for_file: directives_ordering, deprecated_member_use\n'
       '\n'
-      'import \'dart:html\' show Window;\n'
-      'export \'dart:html\' show window, Window;\n'
+      'import \'dart:html\' show Window, Document;\n'
+      'export \'dart:html\' show window, Window, Document;\n'
       '\n'
+      'typedef DocumentOrStubbed = Document;\n'
       'typedef WindowOrStubbed = Window;\n'
       '',
   'site|lib/generated/imports/_vm.dart':
@@ -89,10 +151,11 @@ final importsStubsOutput = {
       '\n'
       '// ignore_for_file: directives_ordering, deprecated_member_use\n'
       '\n'
-      'import \'dart:io\' show HttpServer;\n'
-      'export \'dart:io\' show HttpServer;\n'
+      'import \'dart:io\' show HttpServer, HttpHeaders;\n'
       'export \'dart:collection\' show IterableExtensions;\n'
+      'export \'dart:io\' show HttpServer, HttpHeaders;\n'
       '\n'
+      'typedef HttpHeadersOrStubbed = HttpHeaders;\n'
       'typedef HttpServerOrStubbed = HttpServer;\n'
       '',
   'site|lib/generated/imports/_stubs.dart':
@@ -103,12 +166,6 @@ final importsStubsOutput = {
       '\n'
       '// ignore_for_file: directives_ordering, non_constant_identifier_names\n'
       '\n'
-      'dynamic window;\n'
-      'dynamic Window;\n'
-      'typedef WindowOrStubbed = dynamic;\n'
-      'dynamic HttpServer;\n'
-      'typedef HttpServerOrStubbed = dynamic;\n'
-      '\n'
       'extension IterableExtensions on dynamic {\n'
       '  dynamic get indexed => null;\n'
       '  dynamic get firstOrNull => null;\n'
@@ -116,5 +173,15 @@ final importsStubsOutput = {
       '  dynamic get singleOrNull => null;\n'
       '  dynamic get elementAtOrNull => null;\n'
       '}\n'
+      '\n'
+      'dynamic window;\n'
+      'dynamic Window;\n'
+      'typedef WindowOrStubbed = dynamic;\n'
+      'dynamic Document;\n'
+      'typedef DocumentOrStubbed = dynamic;\n'
+      'dynamic HttpServer;\n'
+      'typedef HttpServerOrStubbed = dynamic;\n'
+      'dynamic HttpHeaders;\n'
+      'typedef HttpHeadersOrStubbed = dynamic;\n'
       '',
 };

